google_guest_agent/snapshot_listener.go

```diff
@@ -0,0 +1,252 @@
+
+	if conf.PreSnapshotScriptURL == "" && conf.PostSnapshotScriptURL == "" {
+		msg := "neither pre or post snapshot script has been configured"
+		logger.Errorf("%s", msg)
+		return conf, &invalidSnapshotConfig{msg}
+	}
+
+	return conf, nil
+}
+
+func runScript(path string, disks string, timeout int) (int, error) {
+	logger.Infof("running script at: %s", path)
+	cmd := exec.Command(path, disks)
+	err := cmd.Start()
+	if err != nil {
+		logger.Errorf("failed to start script at %s: %v", path, err)
+		return 0, err
+	}
+
+	done := make(chan error)
+	go func() { done <- cmd.Wait() }()
+
+	// Start timer
+	timer := time.After(time.Second * time.Duration(timeout))
+
+	select {
+	case <-timer:
+		logger.Errorf("timeout while running script at %s", path)
+		// Exit code of the timeout command when it times out
+		return 124, &scriptsTimeoutError{path}
+	case err := <-done:
+		if err != nil {
+			exiterr := err.(*exec.ExitError)
+			status := exiterr.Sys().(syscall.WaitStatus)
+			logger.Errorf("error while running script at %s, return with exit code %d", path, status.ExitStatus())
+			return status.ExitStatus(), nil
+		}
+	}
+	return 0, nil
+}
+
+func handleRequest(config snapshotConfig, path string, disks string) (pb.AgentErrorCode, int) {
+	ec, err := runScript(path, disks, config.Timeout)
+	if os.IsNotExist(err) {
+		logger.Infof("script not found at %s: ", path)
+		return pb.AgentErrorCode_SCRIPT_NOT_FOUND, 0
+	}
+
+	if _, ok := err.(*scriptsTimeoutError); ok {
+		logger.Infof("script at: %s timed out", path)
+		return pb.AgentErrorCode_SCRIPT_TIMED_OUT, ec
+	}
+
+	if !config.ContinueOnScriptError && ec != 0 {
+		logger.Infof("script at: %s failed with error code %d", path, ec)
+		return pb.AgentErrorCode_UNHANDLED_SCRIPT_ERROR, ec
+	}
+
+	return pb.AgentErrorCode_NO_ERROR, ec
+}
+
+func listenForSnapshotRequests(address string, requestChan chan *pb.GuestMessage) {
+	for {
+		// Start hanging connection on server that feeds to channel
+		logger.Infof("attempting to connect to snapshot service at %s", address)
+		conn, err := grpc.Dial(address, grpc.WithInsecure())
+		if err != nil {
+			logger.Errorf("failed to connect: %v", err)
+			continue
+		}
+
+		c := pb.NewSnapshotServiceClient(conn)
+		ctx, cancel := context.WithCancel(context.Background())
+		guestReady := pb.GuestReady{
+			RequestServerInfo: false,
+		}
+		r, err := c.CreateConnection(ctx, &guestReady)
+		if err != nil {
+			logger.Errorf("error creating connection: %v", err)
+			continue
+		}
+		logger.Infof("created hanging connection with snapshot service")
+		for {
+			request, err := r.Recv()
+			if err != nil {
+				logger.Errorf("error reading request: %v", err)
+				break
+			}
+			logger.Infof("received snapshot request")
+			requestChan <- request
+		}
+		cancel()
+	}
+}
+
+func getSnapshotResponse(guestMessage *pb.GuestMessage) *pb.SnapshotResponse {
+	switch request := guestMessage.Msg.(type) {
```

Comment:
consider a paradigm we use for this kind of test in osconfig: https://github.com/GoogleCloudPlatform/osconfig/blob/master/policies/recipes/installrecipe.go#L89

Reply:
See if this is what you had in mind.